File: src/plugins/claude-code/hooks/post-tool-use.ts
#!/usr/bin/env node
import { BaseHook } from './base';
import { PostToolUseEvent } from '../types';
import { detectEmotion, Emotion } from '../../../tts/index';
import { announceIfEnabled } from '../../../tts/announce';

export class PostToolUseHook extends BaseHook {
  constructor() {
    super('post-tool-use');
  }

  async execute(): Promise<void> {
    const input = await this.readStdin();

    // If no input (e.g., run manually), exit gracefully
    if (!input) {
      return;
    }

    const event = this.parseInput(input) as PostToolUseEvent;

    if (!event) {
      this.logger.warn('Invalid post-tool-use event');
      return;
    }

    // Log the event with performance metrics
    this.logEvent({
      type: 'post-tool-use',
      timestamp: new Date().toISOString(),
      data: {
        ...event,
        duration: event.duration || 0,
        exitCode: event.exitCode || 0,
      },
    });

    // Only announce for significant events
    if (!this.shouldAnnounce(event)) {
      return;
    }

    // Generate message
    const message = this.getStaticMessage(event);
    const emotion = this.determineEmotion(event);

    // Announce the message
    await announceIfEnabled(message, emotion);

    // Track performance metrics
    this.trackPerformance(event);
  }

  private shouldAnnounce(event: PostToolUseEvent): boolean {
    // Announce long-running commands (>5s)
    if (event.duration && event.duration > 5000) return true;

    // Announce failures
    if (event.exitCode !== 0) return true;

    // Announce specific important tools
    const importantTools = ['build', 'test', 'deploy', 'install', 'npm', 'yarn', 'pnpm'];
    if (importantTools.some((tool) => event.tool.toLowerCase().includes(tool))) return true;

    return false;
  }

  private getStaticMessage(event: PostToolUseEvent): string {
    const seconds = Math.round((event.duration || 0) / 1000);
    const toolName = this.getToolDisplayName(event.tool);

    if (event.exitCode === 0) {
      if (seconds > 0) {
        return `${toolName} completed in ${seconds} seconds`;
      }
      return `${toolName} completed`;
    } else {
      return `${toolName} failed with error`;
    }
  }

  private determineEmotion(event: PostToolUseEvent): Emotion {
    if (event.exitCode === 0) {
      // Long tasks get calm emotion
      if (event.duration && event.duration > 10000) {
        return 'calm';
      }
      return 'cheerful';
    } else {
      return 'concerned';
    }
  }

  private getToolDisplayName(tool: string): string {
    const displayNames: Record<string, string> = {
      bash: 'Command',
      shell: 'Shell command',
      read: 'File read',
      write: 'File write',
      search: 'Search',
      grep: 'Grep search',
      find: 'Find',
    };

    return displayNames[tool.toLowerCase()] || tool;
  }

  private async announce(message: string, emotion?: string): Promise<void> {
    try {
      // Use detected emotion if not provided
      const finalEmotion =
        emotion || detectEmotion(message, { success: message.includes('completed') });
      await announceIfEnabled(message, finalEmotion as Emotion);
    } catch (error) {
      this.logger.error(`TTS error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  private trackPerformance(event: PostToolUseEvent): void {
    // Simple performance tracking
    const metrics = {
      tool: event.tool,
      duration: event.duration || 0,
      success: event.exitCode === 0,
      timestamp: new Date().toISOString(),
    };

    this.jsonLogger.info({
      type: 'performance',
      data: metrics,
    });
  }
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  const hook = new PostToolUseHook();
  void hook.run();
}

File: .claude/hooks/post_tool_use.py
#!/usr/bin/env python3

import json
import sys
import subprocess
from pathlib import Path

def get_tts_command():
    """Get the stts command path"""
    return "stts"

def main():
    try:
        # Read JSON from stdin
        data = json.load(sys.stdin)
        
        # Log the event for debugging
        log_dir = Path.home() / ".stts" / "logs"
        log_dir.mkdir(parents=True, exist_ok=True)
        
        # Append to log file
        log_file = log_dir / "post_tool_use.json"
        logs = []
        if log_file.exists():
            try:
                with open(log_file, 'r') as f:
                    logs = json.load(f)
            except:
                logs = []
        
        logs.append(data)
        
        with open(log_file, 'w') as f:
            json.dump(logs, f, indent=2)
        
        # Extract tool info
        tool = data.get('tool', '')
        exit_code = data.get('exitCode', 0)
        duration = data.get('duration', 0)
        
        # Only announce failures or long-running commands
        if exit_code != 0:
            message = f"{tool} failed"
            # Announce via TTS
            try:
                subprocess.run([
                    get_tts_command(), "say", message
                ], capture_output=True, timeout=5)
            except:
                pass  # Fail silently
        elif duration > 5000:  # Over 5 seconds
            seconds = duration / 1000
            message = f"{tool} completed in {seconds:.1f} seconds"
            # Announce via TTS
            try:
                subprocess.run([
                    get_tts_command(), "say", message
                ], capture_output=True, timeout=5)
            except:
                pass  # Fail silently
        
    except Exception:
        pass  # Fail silently
    
    sys.exit(0)

if __name__ == "__main__":
    main()